### ISSUE TITLE
Pro-actively abort pull operations which will fail due to disk space

### ISSUE THREAD

**Issue by uajain** [ostreedev/ostree#1557]
Pull operations which are guaranteed to fail due to unavailability of disk space for the final deployment of the application should be pro-actively aborted. If OSTree has some mechanism so that the size 
 of the remaining download(or pull) can be known , it can be checked periodically against the disk space availability (using fstatfs). This also avoids the situation where the user is left with disk-full(due to partial pull) and there is no successful deployment of the application.

**Comment by cgwalters** [comment]
I assume these are flatpak pulls, and not host pulls?  See also https://github.com/flatpak/flatpak/issues/975 

On EOS is the space check omitted for both since it's a unirepo?

There are some tricky aspects to this.  One is around simply knowing the size.  With static deltas, we do.  Without that...there's some work on size metadata in the commit object, but I don't think we ended up using that as is?  It's temping to revisit.

Another possibly simple thing to do is if we actually hit `ENOSPC`, to unwind all repository temporary space.

**Comment by mwleeds** [comment]
`min-free-space-percent=0` is in the config of the repo used by both flatpak and eos-updater, if that answers your question

**Comment by ramcq** [comment]
> There are some tricky aspects to this. One is around simply knowing the size. With static deltas, we do. Without that...there's some work on size metadata in the commit object, but I don't think we ended up using that as is? It's temping to revisit.

Yeah, Endless has the object size metadata in our OS repo (for host updates), but it's not available on other remotes that Endless uses such as Flathub. It would allow us to get the expected operation size much more accurately, but that doesn't necessarily matter - because during an operation, the available disk space can also change.

On non-delta pulls, I was imagining that in the progress callback (given that the total size can grow over time as we fetch the object tree) we even actually check the expected additional disk usage vs fstatfs free space periodically during the pull operation, and stop if they are not going to converge.

> Another possibly simple thing to do is if we actually hit ENOSPC, to unwind all repository temporary space.

We should do this as well - basically roll back a cached pull if it fills the disk, rather than leaving the disk totally full. But this isn't the best thing to do for the user: disk space can be scarce, but bandwidth can be expensive. The best thing to do is to waste neither of them, hence pro-actively predicting the out of disk and avoiding continuing the download in this case.
